Replace Win32 PAL implementation with minipal implementation

AaronRobinsonMSFT opened the pull request:
This PR removes the PAL Win32 emulation of `CRITICAL_SECTION`. The emulation logic is unnecessarily complicated and used both a pthread mutex _and_ condition variable. For CoreCLR purposes a mutex is sufficient.

A new minipal API was introduced, `minipal_critsect`. This uses the Win32 `CRITICAL_SECTION` on the Windows platform and `pthread_mutex_t` on all other platforms. This implementation aligns with a private "critical section" defined for the GC and native AOT. All now use this new API.

The `CRITICAL_SECTION` associated APIs and tests in the CoreCLR PAL have been removed.

AaronRobinsonMSFT committed:
Reduce critsect.h header usage
Remove PALCEnterCriticalSection and PALCLeaveCriticalSection

AaronRobinsonMSFT committed:
Remove typedef options for CRITICAL_SECTION

AaronRobinsonMSFT committed:
Create PAL critical section
Delete unused debugging APIs in PAL CS code

AaronRobinsonMSFT committed:
Convert C++ API to C

AaronRobinsonMSFT committed:
Remove unused ThreadStore::CanAcquireLock().

AaronRobinsonMSFT committed:
Remove critsec API from coreclr/minipal

AaronRobinsonMSFT committed:
Add critical section API to native/minipal.

AaronRobinsonMSFT committed:
WIP

AaronRobinsonMSFT committed:
Rename critsec.h to critsect.h
Get native AOT to build.

AaronRobinsonMSFT committed:
Consistent naming

AaronRobinsonMSFT committed:
Build on Windows

AaronRobinsonMSFT committed:
Fix macOS build

AaronRobinsonMSFT committed:
Changes to PAL

AaronRobinsonMSFT committed:
Non-windows WIP

AaronRobinsonMSFT committed:
WIP

AaronRobinsonMSFT committed:
Fix build on macos

AaronRobinsonMSFT committed:
WIP

AaronRobinsonMSFT committed:
DAC fixes

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into replace_win32_critsec

AaronRobinsonMSFT committed:
Unused local

huoyaoyuan commented:
FWIW, I tested `std::mutex` and `mtx_t` on Windows. `std::mutex` is about the same speed of `CRITICAL_SECTION`, and `mtx_t` is 2x slower. We are unable to use `std::mutex` because it's not noexcept.

huoyaoyuan commented on src/native/minipal/critsect.c:
```diff
@@ -0,0 +1,68 @@
+#endif // _DEBUG
+}
+
+void minipal_critsect_enter(minipal_critsect* cs)
```
Based on my testing, adding a NOINLINE caller has 10% performance impact on EnterCriticalSection/LeaveCriticalSection pair. Is it worthy to move the enter/leave implementations to header? Or would LTCG take care of this?

AaronRobinsonMSFT replied:
> Based on my testing, adding a NOINLINE caller has 10% performance impact on EnterCriticalSection/LeaveCriticalSection pair. 

Please share how you determined this.

huoyaoyuan replied:
It appears to be a test error. Further runs don't support this and sometimes direct call gets slower.

Code:
<details>

```cpp
DECLSPEC_NOINLINE void EnterCriticalSection_NoInline(LPCRITICAL_SECTION lpcs)
{
	EnterCriticalSection(lpcs);
}

DECLSPEC_NOINLINE void LeaveCriticalSection_NoInline(LPCRITICAL_SECTION lpcs)
{
	LeaveCriticalSection(lpcs);
}

int main()
{
	LARGE_INTEGER start, end;
	{
		CRITICAL_SECTION cs;
		InitializeCriticalSection(&cs);

		QueryPerformanceCounter(&start);

		for (int i = 0; i < 100000; i++)
		{
			EnterCriticalSection(&cs);
			LeaveCriticalSection(&cs);
		}

		QueryPerformanceCounter(&end);
		DeleteCriticalSection(&cs);

		cout << "CriticalSection " << end.QuadPart - start.QuadPart << endl;
	}
	
	{
		CRITICAL_SECTION cs;
		InitializeCriticalSection(&cs);

		QueryPerformanceCounter(&start);

		for (int i = 0; i < 100000; i++)
		{
			EnterCriticalSection_NoInline(&cs);
			LeaveCriticalSection_NoInline(&cs);
		}

		QueryPerformanceCounter(&end);
		DeleteCriticalSection(&cs);

		cout << "CriticalSection(NoInline) " << end.QuadPart - start.QuadPart << endl;
	}
}
```
</details>

jkoritzinsky commented on src/native/minipal/critsect.h:
```diff
@@ -0,0 +1,68 @@
+#endif // __cplusplus
+
+#ifdef __cplusplus
+class MinipalCritSectHolder final
```
Add a `minipal` namespace instead of a class name prefix?

AaronRobinsonMSFT replied:
Good idea. I like it.

jkoritzinsky commented on src/native/minipal/critsect.h:
```diff
@@ -0,0 +1,68 @@
+    minipal_critsect* _cs;
+
+public:
+    explicit MinipalCritSectHolder(minipal_critsect* cs)
```
I know it's less common in our code-base, but could we use references here instead of pointers? I think the usage would look cleaner.

AaronRobinsonMSFT replied:
Sure.

AaronRobinsonMSFT committed:
Remove PAL tests for CRITICAL_SECTION
Remove any non-running tests requiring CRITICAL_SECTION
Remove unused PAL header referencing CRITICAL_SECTION (internal.h)

AaronRobinsonMSFT commented:
> FWIW, I tested `std::mutex` and `mtx_t` on Windows. `std::mutex` is about the same speed of `CRITICAL_SECTION`, and `mtx_t` is 2x slower. We are unable to use `std::mutex` because it's not noexcept.

It isn't just because of `noexcept`. On Windows `std::mutex` will pull in the concrt, which is a huge.

> `mtx_t`

clang on macOS doesn't support C11 threads. It isn't an option in any case.

AaronRobinsonMSFT committed:
Feedback and jit format

AaronRobinsonMSFT committed:
Increase linux 64-bit max size.

am11 commented:
If the implementation is shared between:
* coreclr and nativeaot, we put it under `src/coreclr/minipal`.
* coreclr/natvieaot and libs/corehost/mono, we put it under `src/native/minipal`.

This PR is mixing the existing pattern.

jkotas commented:
> If the implementation is shared between:
> coreclr and nativeaot, we put it under src/coreclr/minipal.
> coreclr/natvieaot and libs/corehost/mono, we put it under src/native/minipal.

`src/native/minipal` have been used to abstract general basic services. Simple lock fits the bill. It is not necessary for everything to be switched to the minipal implementation right way. For example, the initial implementation of the timers was like that.

`src/coreclr/minipal` have been used for functionality that's very specific and not a general basic services, like the double mapper.

am11 commented:
> It is not necessary for everything to be switched to the minipal implementation right way.

If there is a plan to share this impl. with other subsets, then it makes sense. Otherwise, choosing among `src/coreclr/runtime`, `src/coreclr/minipal` and `src/native/minipal` to share code between nativeaot and coreclr is getting vague and not how they were initially defined.

jkotas commented on src/coreclr/inc/crosscomp.h:
```diff
@@ -705,77 +706,79 @@ typedef struct _KNONVOLATILE_CONTEXT_POINTERS_EX
+// We have this requirement for cross OS compiling the DAC.
+struct tgt_minipal_critsect final
+{
+    minipal_critsect _cs;
+
+    // This is unused padding to ensure struct size.
+    BYTE _dacPadding[DAC_CS_MAX_SIZE - sizeof(_cs)];
```
```suggestion
    union
    {
        minipal_critsect _cs;

        // Padding to ensure struct size.
        char _dacPadding[DAC_CS_MAX_SIZE];
        
        // Ensure alignment
        void* _dacAlignment;
    }
```

jkoritzinsky replied:
Could we use `alignas(alignof(void*))` instead of including it for alignment requirements?

AaronRobinsonMSFT replied:
It isn't required. The current PAL needed it because the space being allocated was a placeholder for a type defined in the PAL with specific platform implementation details. There was then a C-style cast from the PAL definition of `CRITICAL_SECTION` to the internal implementation of `PAL_CRITICAL_SECTION`. Forcing that alignment was needed to ensure the `BYTE[]` was properly aligned to what was in `PAL_CRITICAL_SECTION`.

jkotas replied:
With the current state of the PR, we depend on alignment requirement for the OS-specific `minipal_critsect` to be same (ie. `sizeof(void*)`) across platforms. Is it possible to assert it somewhere?

jkotas replied:
The scenario that I am concerned about are Windows-hosted Linux-targeting DACs. We are overlaying Windows CRITICAL_SECTION over Linux pthread_mutex on that case that are completely different structures.

A technically more correct definition for DAC builds that avoids this questionable overlaying would look something like this:
```
    union
    {
        // DAC builds want to have the data layout of the target system. Make sure that the host minipal_critsect
        // does not influence the target data layout 
#infdef DACCESS_COMPILE
        minipal_critsect _cs;
#endif

        // Padding to ensure struct size.
        char _dacPadding[DAC_CS_MAX_SIZE];
        
        // Ensure alignment
        void* _dacAlignment;
    }
```

AaronRobinsonMSFT replied:
Ah. I see your concern now. Sure, I can use a similar definition.

jkotas commented on src/coreclr/nativeaot/Runtime/EHHelpers.cpp:
```diff
@@ -490,7 +487,7 @@ LONG WINAPI RhpVectoredExceptionHandler(PEXCEPTION_POINTERS pExPtrs)
             // Do not use ASSERT_UNCONDITIONALLY here. It will crash because of it consumes too much stack.
             PalPrintFatalError("\nProcess is terminating due to StackOverflowException.\n");
-            RaiseFailFastException(pExPtrs->ExceptionRecord, pExPtrs->ContextRecord, 0);
+            PalRaiseFailFastException(pExPtrs->ExceptionRecord, pExPtrs->ContextRecord, 0);
```
This is Windows-specific code calling Windows-specific API. There is no point going through a PAL layer in cases like these.

AaronRobinsonMSFT replied:
I see. The problem with this model is there is an odd consistency in files. It isn't obvious/clear on when we can avoid the PAL. If this file was clearly in a "windows" directory it would be obvious. In thics case though we have 100s of lines of code that defined in complex ways that make it tricky to know. If we instead just implement a PAL consistently we can use the simple rule of "only use PAL APIs". I'll change it, but I don't think this is a helpful way of writing code in a file that isn't defined as being platform specific.

See the other "Pal" functions in the same block.

AaronRobinsonMSFT commented on src/coreclr/inc/crosscomp.h:
```diff
@@ -705,77 +706,79 @@ typedef struct _KNONVOLATILE_CONTEXT_POINTERS_EX
-};
+#else // !DACCESS_COMPILE
+
+#if defined(__APPLE__)
```
I guess we really don't. I was trying to keep the system similar and stepping back it seems this was unnecessary.

AaronRobinsonMSFT committed:
Feedback and build breaks.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into replace_win32_critsec

AaronRobinsonMSFT committed:
Update Windows max size.

AaronRobinsonMSFT committed:
Fix up native AOT PAL for unix

AaronRobinsonMSFT committed:
FIx enter call that was accidentally made leave.
Remove unnecessary getthreads in PAL.

AaronRobinsonMSFT commented:
Given the current progress of the CI, I think this is ready to be reviewed.

huoyaoyuan commented on src/coreclr/tools/superpmi/superpmi-shared/logging.h:
```diff
@@ -65,7 +67,7 @@ class Logger
     static HANDLE           s_logFile;
     static char*            s_logFilePath;
-    static CRITICAL_SECTION s_critSec;
+    static minipal_critsect      s_critSec;
```
Nit: format

huoyaoyuan commented on src/coreclr/tools/superpmi/superpmi-shared/logging.cpp:
```diff
@@ -20,7 +20,7 @@ bool             Logger::s_initialized = false;
 HANDLE           Logger::s_logFile     = INVALID_HANDLE_VALUE;
 char*            Logger::s_logFilePath = nullptr;
-CRITICAL_SECTION Logger::s_critSec;
+minipal_critsect      Logger::s_critSec;
```
Nit: format

huoyaoyuan commented on src/coreclr/dlls/mscordbi/CMakeLists.txt:
```diff
@@ -75,6 +75,7 @@ set(COREDBI_LIBRARIES
     mdruntimerw-dbi
     mddatasource_dbi
     corguids
+    minipal_objects
```
Why not linking `minipal`?

AaronRobinsonMSFT replied:
Seems like a mistake to me. Will change.

AaronRobinsonMSFT committed:
Feedback

jkotas commented on src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.cpp:
```diff
@@ -17,6 +17,7 @@
 #include <unistd.h>
 #endif
 
+#include <minipal/critsect.h>
```
```suggestion
```
Is this needed?

jkotas commented on src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.h:
```diff
@@ -830,7 +830,7 @@ uint32_t
 {
     STATIC_CONTRACT_NOTHROW;
-#ifdef _INC_WINDOWS
+#ifndef TARGET_UNIX
```
```suggestion
#ifdef HOST_WINDOWS
```

jkotas commented on src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.h:
```diff
@@ -882,7 +882,7 @@ ep_rt_system_time_get (EventPipeSystemTime *system_time)
     STATIC_CONTRACT_NOTHROW;
 
-#ifdef _INC_WINDOWS
+#ifndef TARGET_UNIX
```
```suggestion
#ifdef HOST_WINDOWS
```

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:
```diff
@@ -1038,3 +1036,63 @@ void SetSSP(CONTEXT *pContext, uintptr_t ssp)
+
+void PalRaiseFailFastException(PEXCEPTION_RECORD arg1, PCONTEXT arg2, uint32_t arg3)
+{
+    ::RaiseFailFastException(arg1, arg2, arg3);
```
This will degrade the diagnostics with `FAIL_FAST_GENERATE_EXCEPTION_ADDRESS` flag. We won't necessarily get the actual stop where the FailFast was called with this wrapper.

jkotas commented on src/coreclr/nativeaot/Runtime/rhassert.h:
```diff
@@ -62,6 +62,6 @@ void Assert(const char * expr, const char * file, unsigned int line_num, const c
 #define FAIL_FAST_GENERATE_EXCEPTION_ADDRESS 0x1
 
-#define RhFailFast() RaiseFailFastException(NULL, NULL, FAIL_FAST_GENERATE_EXCEPTION_ADDRESS)
+#define RhFailFast() PalRaiseFailFastException(NULL, NULL, FAIL_FAST_GENERATE_EXCEPTION_ADDRESS)
```
```suggestion
#ifdef HOST_WINDOWS
#define RhFailFast() RaiseFailFastException(NULL, NULL, FAIL_FAST_GENERATE_EXCEPTION_ADDRESS)
#else
void RhFailFast();
#endif
```
We should add `RhFailFast` as the PAL API (it needs to be macro to make FAIL_FAST_GENERATE_EXCEPTION_ADDRESS work correctly - see my other comment), and delete  PalRaiseFailFastException/RaiseFailFastException and relate goo from the PAL.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into replace_win32_critsec

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:
```diff
@@ -1038,3 +1036,63 @@ void SetSSP(CONTEXT *pContext, uintptr_t ssp)
+    return static_cast<uint32_t>(::GetCurrentProcessId());
+}
+
+#ifdef UNICODE
+uint32_t PalGetEnvironmentVariable(_In_opt_ LPCWSTR lpName, _Out_writes_to_opt_(nSize, return + 1) LPWSTR lpBuffer, _In_ uint32_t nSize)
+{
+    return ::GetEnvironmentVariableW(lpName, lpBuffer, nSize);
+}
+#else
+uint32_t PalGetEnvironmentVariable(_In_opt_ LPCSTR lpName, _Out_writes_to_opt_(nSize, return + 1) LPSTR lpBuffer, _In_ uint32_t nSize)
+{
+    return ::GetEnvironmentVariableA(lpName, lpBuffer, nSize);
+}
+#endif
```
```suggestion
uint32_t PalGetEnvironmentVariable(_In_opt_ LPCWSTR lpName, _Out_writes_to_opt_(nSize, return + 1) LPWSTR lpBuffer, _In_ uint32_t nSize)
{
    return ::GetEnvironmentVariableW(lpName, lpBuffer, nSize);
}
```
We are always building with UNICODE defined on Windows.

AaronRobinsonMSFT commented on src/coreclr/inc/crosscomp.h:
```diff
@@ -705,77 +706,36 @@ typedef struct _KNONVOLATILE_CONTEXT_POINTERS_EX
+#elif defined(TARGET_LINUX) || defined(TARGET_ANDROID)
+#define DAC_CS_MAX_SIZE 48
+#elif defined(TARGET_WINDOWS)
+#define DAC_CS_MAX_SIZE 64
```
No. It was failing on *nix. We are already saving substantial here and I would instead prefer to get the wins when we move to the cDAC rather than making this section more complicated.

jkotas replied:
Why/how was it failing on Unix?

AaronRobinsonMSFT replied:
I think this might have been because of the missing `CROSS_COMPILE` block on the static_assert. I'll confirm.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into replace_win32_critsec

AaronRobinsonMSFT committed:
Feedback - first pass

AaronRobinsonMSFT committed:
Add back __stdcall

AaronRobinsonMSFT commented on src/coreclr/nativeaot/Runtime/PalRedhawk.h:
```diff
@@ -50,6 +52,7 @@
 #else   // !defined(__i386__)
 
 #define __cdecl
+#define __stdcall
```
There are several issues here. @#$%@#$

AaronRobinsonMSFT committed:
Remove __stdcall usage

AaronRobinsonMSFT committed:
Feedback - second pass.

AaronRobinsonMSFT committed:
Fix cross build

davidwrighton commented on src/coreclr/nativeaot/Runtime/rhassert.h:
```diff
@@ -60,8 +60,12 @@ void Assert(const char * expr, const char * file, unsigned int line_num, const c
     ASSUME(0);  \
 
-#define FAIL_FAST_GENERATE_EXCEPTION_ADDRESS 0x1
+//#define FAIL_FAST_GENERATE_EXCEPTION_ADDRESS 0x1
```
Why is this commented out?

AaronRobinsonMSFT replied:
That should be removed. I was seeing if I could remove it. Thanks!

AaronRobinsonMSFT committed:
Fix redefinition error.

AaronRobinsonMSFT committed:
Remove unnecessary define

jkotas commented on src/coreclr/inc/crosscomp.h:
```diff
@@ -705,77 +714,38 @@ typedef struct _KNONVOLATILE_CONTEXT_POINTERS_EX
+#elif defined(TARGET_LINUX) || defined(TARGET_ANDROID)
+#define DAC_CS_MAX_SIZE 64
+#elif defined(TARGET_WINDOWS)
+#define DAC_CS_MAX_SIZE 40
```
```suggestion
:#ifdef TARGET_64BIT
#define DAC_CS_MAX_SIZE 40
#else
#define DAC_CS_MAX_SIZE 24
#endif
```
Avoid regressing wn-x86

AaronRobinsonMSFT committed:
Special case 32-bit Windows for max size

AaronRobinsonMSFT commented on src/native/minipal/critsect.h:
```diff
@@ -0,0 +1,71 @@
+    MINIPAL_CRITSECT_IMPL _impl;
+} minipal_critsect;
+
+// Initialize the critical section
```
That is a fair argument. I'm fine using mutex actually, but should we then also consider calling it "recursive_mutex"?

jkotas replied:
Either would be fine with me.

AaronRobinsonMSFT committed:
Change many names to mutex from critical section.

AaronRobinsonMSFT committed:
Fix format in jit.

jkotas committed:
Merge branch 'main' into replace_win32_critsec
